Add interfaces and mocks

katejefferson opened the pull request:
This PR is gigantic. Sorry. All tests pass, if it is any consolation. 
* Added interfaces for each service 
* Renamed `recurly` package `api` and used `recurly` for the root package 
* Moved custom types to root package 
* Moved `types` into it's own package (null types, dates and times etc) 
* Added mocks for the `TransactionsService` and the `SubscriptionsService`

cristiangraz commented on subscriptions.go:
```diff
@@ -0,0 +1,197 @@
+	}
+)
+
+const (
```
Could you move these up just below the imports? Same comment throughout the PR.

katejefferson replied:
Fixed in [4687d1](https://github.com/blacklightcms/go-recurly/pull/10/commits/9278698024ab9b18b0b4c965637f58f60a4687d1)

cristiangraz commented on api/client.go:
```diff
@@ -30,20 +32,17 @@ type (
-		AddOns        *AddOnsService
-		Subscriptions *SubscriptionsService
-		Transactions  *TransactionsService
+		Accounts      recurly.AccountsService
```
Can you move this `Client` struct, as well as the two functions (`newRequest` and `do`) into the root package, and have `NewClient` below return `*recurly.Client`? Youll need to be `NewRequest` and `Do` exported methods and update each of the services. Beyond that, I don't see anything that would cause import cycle issues in the `Client` or it's two functions.

katejefferson replied:
Fixed in [b74efc](https://github.com/blacklightcms/go-recurly/pull/10/commits/9bea2c2253c260f6d17ac6dd436edd7566b74efc)

katejefferson commented on mocks/mocks.go:
```diff
@@ -0,0 +1,92 @@
+var _ recurly.Transactions = &MockTransactionsService{}
+
+// MockTransactionsService mocks the transaction service.
+type MockTransactionsService struct {
```
Fixed in [bbaea8](https://github.com/blacklightcms/go-recurly/pull/10/commits/fad5453852cf15bd32cdadb74bfe5677a4bbaea8)

katejefferson commented on accounts.go:
```diff
@@ -0,0 +1,109 @@
+	"github.com/blacklightcms/go-recurly/types"
+)
+
+type (
```
Fixed in [b83c87](https://github.com/blacklightcms/go-recurly/pull/10/commits/e775de18bcd2e4a91905e2431ef00d2d33b83c87)

katejefferson commented:
* Moved `types` into recurly package in [d337c3](https://github.com/blacklightcms/go-recurly/pull/10/commits/e0b0b0e8812b1e5878e749f82ffb5570c5d337c3)
* Updated yaml in [f04883](https://github.com/blacklightcms/go-recurly/pull/10/commits/8fe5bdc9d12449c43e0faec44757b9a517f04883)
* Moved `webhooks` to same level as `api` in [9a10a3](https://github.com/blacklightcms/go-recurly/pull/10/commits/5b0680471b2b340fc9e92ec4814310ecec9a10a3)
* Renamed `mocks` to `mock` in [bbaea8](https://github.com/blacklightcms/go-recurly/pull/10/commits/fad5453852cf15bd32cdadb74bfe5677a4bbaea8)

katejefferson commented on invoices.go:
```diff
@@ -0,0 +1,121 @@
+	}
+)
+
+const (
```
Fixed in [56f7a0](https://github.com/blacklightcms/go-recurly/pull/10/commits/70680f140c93eac4f28ac08599257fe42756f7a0)

katejefferson commented on api/accounts.go:
```diff
@@ -0,0 +1,137 @@
+	client *Client
+}
+
+const (
```
Fixed in [56f7a0](https://github.com/blacklightcms/go-recurly/pull/10/commits/70680f140c93eac4f28ac08599257fe42756f7a0)

katejefferson commented on mocks/mocks.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+// MockClient mocks the recurly client.
+type MockClient struct {
```
Fixed in [bbaea8](https://github.com/blacklightcms/go-recurly/pull/10/commits/fad5453852cf15bd32cdadb74bfe5677a4bbaea8)

cristiangraz commented on client.go:
```diff
@@ -10,71 +10,64 @@ import (
-	c := &Client{
-		client:    httpClient,
+	client := &Client{
+		Client:    httpClient,
```
Can we keep the `*http.Client` unexported?

katejefferson replied:
Unexported in [5238e2](https://github.com/blacklightcms/go-recurly/pull/10/commits/322b61900754b8bcf3719fcf55a8b1a7955238e2)

cristiangraz commented on client.go:
```diff
@@ -10,71 +10,64 @@ import (
-// newRequest creates an authenticated API request that is ready to send.
-func (c *Client) newRequest(method string, action string, params Params, body interface{}) (*http.Request, error) {
+// NewRequest creates an authenticated API request that is ready to send.
+func (c *Client) NewRequest(method string, action string, params Params, body interface{}) (*http.Request, error) {
```
Can we revert this back to unexported?

katejefferson replied:
Unexported in [5238e2](https://github.com/blacklightcms/go-recurly/pull/10/commits/322b61900754b8bcf3719fcf55a8b1a7955238e2)

cristiangraz commented on client.go:
```diff
@@ -108,14 +101,14 @@ func (c *Client) newRequest(method string, action string, params Params, body in
 // It returns a Response object that provides a wrapper around http.Response
 // with some convenience methods.
-func (c *Client) do(req *http.Request, v interface{}) (*Response, error) {
+func (c *Client) Do(req *http.Request, v interface{}) (*Response, error) {
```
Can we revert this back to unexported?

katejefferson replied:
Unexported in [5238e2](https://github.com/blacklightcms/go-recurly/pull/10/commits/322b61900754b8bcf3719fcf55a8b1a7955238e2)

cristiangraz commented on mock/client.go:
```diff
@@ -0,0 +1,21 @@
+
+	client.Subscriptions = &SubscriptionsService{}
+	client.Transactions = &TransactionsService{}
+
```
Can you set all of the other services without mocks to nil so they panic until we implement them? Otherwise someone using this mock client will have a mix of mocks and actual api implementations.

katejefferson replied:
Added in [5238e2](https://github.com/blacklightcms/go-recurly/pull/10/commits/322b61900754b8bcf3719fcf55a8b1a7955238e2)
